### PR TITLE
Add inference section to task guides

### PR DESCRIPTION
**Pull request by stevhliu**
Currently, the task guides only show how to finetune a model, but it doesn't directly connect the dots to how you can use that model for inference. For completeness, this PR adds a section to the task guides to show how to use a model for inference after finetuning. This gives users a better overview of the model lifecycle.

In doing so, when we update `task_summary.mdx`, we can focus less on the practical steps of how to use a model for inference (we can add links to the task guides) and instead discuss the more theoretical aspects of these tasks as intended by the Conceptual Guide section.

### CODE REVIEW

File: docs/source/en/tasks/language_modeling.mdx [hunk 1]
```diff
@@ -303,6 +303,71 @@ Call [`fit`](https://keras.io/api/models/model_training_apis/#fit-method) to fin
...
+```
+</pt>
+<tf>
+Tokenize the text and return the `input_ids` as TensorFlow tensors (you can use the model you just finetuned or a pretrained model for causal language modeling):
+
+```py
+>>> from transformers import AutoTokenizer
+
+>>> tokenizer = AutoTokenizer.from_pretrained("distilgpt2")
+>>> inputs = tokenizer(prompt, return_tensors="pt").input_ids
```
Comment (sgugger) [inline docs/source/en/tasks/language_modeling.mdx:350]:
```suggestion
>>> inputs = tokenizer(prompt, return_tensors="tf").input_ids
```

File: docs/source/en/tasks/asr.mdx [hunk 2]
```diff
@@ -223,4 +223,52 @@ At this point, only three steps remain:
...
+```py
+>>> from datasets import load_dataset, Audio
+
+>>> dataset = load_dataset("PolyAI/minds14", "en-US", split="train")
+>>> dataset = dataset.cast_column("audio", Audio(sampling_rate=16000))
+>>> sampling_rate = dataset.features["audio"].sampling_rate
+```
+
+<frameworkcontent>
+<pt>
```
Comment (sgugger) [inline docs/source/en/tasks/asr.mdx:243]: Why not use the pipeline function, at least at first? Then dive into processing + model.

File: docs/source/en/tasks/asr.mdx [hunk 3]
```diff
@@ -223,4 +223,52 @@ At this point, only three steps remain:
...
+>>> processor = AutoProcessor.from_pretrained("facebook/wav2vec2-base-960h")
+>>> inputs = processor(dataset[0]["audio"]["array"], sampling_rate=sampling_rate, return_tensors="pt")
+```
+
+Pass your inputs to the model and return the logits:
+
+```py
+>>> from transformers import AutoModelForCTC
+
+>>> model = AutoModelForCTC.from_pretrained("facebook/wav2vec2-base-960h")
```
Comment (sgugger) [inline docs/source/en/tasks/asr.mdx:258]: This needs to be the fine-tuned model otherwise it defeats the point of this section.

File: docs/source/en/tasks/language_modeling.mdx [hunk 4]
```diff
@@ -303,6 +303,71 @@ Call [`fit`](https://keras.io/api/models/model_training_apis/#fit-method) to fin
...
+```
+
+Use the [`~transformers.generation_tf_utils.TFGenerationMixin`] method to generate text. Check out the [Generation](./main_classes/text_generation) to learn more about the different text generation strategies, and parameters for controlling generation.
+
+```py
+>>> from transformers import AutoModelForCausalLM
```
Comment (sgugger) [inline docs/source/en/tasks/language_modeling.mdx:356]:
```suggestion
>>> from transformers import TFAutoModelForCausalLM
```

**Review by sgugger — COMMENTED** [review]
Thanks for this PR, but I'm not sure the inference section is really useful if it doesn't use the model that was finetuned, as people can find how to use a given pretrained model in the Hub directly.

Also, it would be great to show how to use the fine-tuned model in a pipeline first, which is end-to-end, then dive into the details. I left comments on the first task, but this applies to all of them.

Finally, the TensorFlow content is not TensorFlow: it still uses the same code as PyTorch.

**Comment by stevhliu** [comment]
Thanks for the suggestions! I reworked the `sequence_classification` task a bit, and if we like the changes, then I can apply them to the other tasks. Main changes are:

- Encourage users to login with their HF accounts so they can push their finetuned models. Along these lines, I've added `push_to_hub` and the `PushToHubCallback`.
- Added a section to include a `compute_metrics` function so users can evaluate their models during training.

I think adding these two will help the task guides be more complete :)

File: docs/source/en/tasks/sequence_classification.mdx [hunk 5]
```diff
@@ -197,18 +251,163 @@ Configure the model for training with [`compile`](https://keras.io/api/models/mo
...
+```py
+>>> text = "This was a masterpiece. Not completely faithful to the books, but enthralling from beginning to end. Might be my favorite of the three."
+```
+
+The simplest way to try out your finetuned model for inference is to use it in a [`pipeline`]. Map the expected id to their labels with `id2label` and `label2id`:
+
+```py
+>>> id2label = {0: "NEGATIVE", 1: "POSITIVE"}
+>>> label2id = {"NEGATIVE": 0, "POSITIVE": 1}
+```
```
Comment (sgugger) [inline docs/source/en/tasks/sequence_classification.mdx:314]: This should be done before training saving/pushing to the Hub ideally. So the saved/uploaded model is correct.

File: docs/source/en/tasks/sequence_classification.mdx [hunk 6]
```diff
@@ -197,18 +251,163 @@ Configure the model for training with [`compile`](https://keras.io/api/models/mo
...
+>>> tokenizer = AutoTokenizer.from_pretrained("stevhliu/my_awesome_model")
+>>> model = AutoModelForSequenceClassification.from_pretrained(
+...     "stevhliu/my_awesome_model", id2label=id2label, label2id=label2id
+... )
+```
+
+Instantiate the `pipeline` for sentiment analysis with your model and tokenizer, and pass your text to it:
+
+```py
+>>> classifier = pipeline("sentiment-analysis", model=model, tokenizer=tokenizer)
```
Comment (sgugger) [inline docs/source/en/tasks/sequence_classification.mdx:330]:
```suggestion
>>> classifier = pipeline("sentiment-analysis", model="stevhliu/my_awesome_model")
```
If labels are handled previously, you'll be able to do just this, which is way shorted (and framewok-agnostic).

**Review by sgugger — COMMENTED** [review]
Thanks! I think this is way better this way. Just have a couple of comments on the way labels are handled but otherwise changes are good to push to the other tasks!

File: docs/source/en/tasks/translation.mdx [hunk 7]
```diff
@@ -14,30 +14,47 @@ specific language governing permissions and limitations under the License.
 
 <Youtube id="1JvfrvZgi6c"/>
 
-Translation converts a sequence of text from one language to another. It is one of several tasks you can formulate as a sequence-to-sequence problem, a powerful framework that extends to vision and audio tasks. 
+Translation converts a sequence of text from one language to another. It is one of several tasks you can formulate as a sequence-to-sequence problem, a powerful framework for returning some output from an input like translation or summarization. Companies like Google use sequence-to-sequence models to power its products like Translate and Docs.
```
Comment (sgugger) [inline docs/source/en/tasks/translation.mdx:17]: Why are we doing free advertisement for Google here?

File: docs/source/en/tasks/translation.mdx [hunk 8]
```diff
@@ -82,84 +99,113 @@ The preprocessing function needs to:
...
 ```
 
-Use 🤗 Datasets [`~datasets.Dataset.map`] function to apply the preprocessing function over the entire dataset. You can speed up the `map` function by setting `batched=True` to process multiple elements of the dataset at once:
+To apply the preprocessing function over the entire dataset, use 🤗 Datasets [`~datasets.Dataset.map`] method. You can speed up the `map` function by setting `batched=True` to process multiple elements of the dataset at once:
 
 ```py
 >>> tokenized_books = books.map(preprocess_function, batched=True)
 ```
 
+Now create a batch of examples using [`DataCollatorForSeq2Seq`]. It will also *dynamically pad* your text and labels to the length of the longest element in its batch (instead of the entire dataset) so they are a uniform length. While it is possible to pad your text in the `tokenizer` function by setting `padding=True`, dynamic padding is more efficient.
```
Comment (sgugger) [inline docs/source/en/tasks/translation.mdx:108]: The second sentence is not accurate. It's more efficient to dynamically pad the sentences to the longest length in the batch during collation, rather than padding the whole dataset to the maximum length once.

File: docs/source/en/tasks/translation.mdx [hunk 9]
```diff
@@ -169,13 +215,41 @@ At this point, only three steps remain:
...
-To fine-tune a model in TensorFlow, start by converting your datasets to the `tf.data.Dataset` format with [`~TFPreTrainedModel.prepare_tf_dataset`].
+<Tip>
+
+If you aren't familiar with finetuning a model with Keras, take a look at the basic tutorial [here](../training#train-a-tensorflow-model-with-keras)!
+
+</Tip>
+To finetune a model in TensorFlow, start by setting up an optimizer function, learning rate schedule, and some training hyperparameters:
+
+```py
+>>> from transformers import create_optimizer, AdamWeightDecay
```
Comment (sgugger) [inline docs/source/en/tasks/translation.mdx:239]: It doesn't look like the `create_optimizer` function is used.

File: docs/source/en/tasks/token_classification.mdx [hunk 10]
```diff
@@ -126,13 +145,13 @@ Here is how you can create a function to realign the tokens and labels, and trun
...
+To apply the preprocessing function over the entire dataset, use 🤗 Datasets [`~datasets.Dataset.map`] function. You can speed up the `map` function by setting `batched=True` to process multiple elements of the dataset at once:
 
 ```py
 >>> tokenized_wnut = wnut.map(tokenize_and_align_labels, batched=True)
 ```
 
-Use [`DataCollatorForTokenClassification`] to create a batch of examples. It will also *dynamically pad* your text and labels to the length of the longest element in its batch, so they are a uniform length. While it is possible to pad your text in the `tokenizer` function by setting `padding=True`, dynamic padding is more efficient.
+Now create a batch of examples using [`DataCollatorWithPadding`]. It'll also *dynamically pad* your text and labels to the length of the longest element in its batch (instead of the entire dataset) so they are a uniform length. While it is possible to pad your text in the `tokenizer` function by setting `padding=True`, dynamic padding is more efficient.
```
Comment (sgugger) [inline docs/source/en/tasks/token_classification.mdx:154]: AS mentioned in the translation guide, this is a bit misleading. It's more efficient to pad dynamically to the max length in the batch vs padding to the max length of the dataset, but in both cases this uses the `tokenizer` pad method behind the scenes.

File: docs/source/en/tasks/token_classification.mdx [hunk 11]
```diff
@@ -151,39 +170,120 @@ Use [`DataCollatorForTokenClassification`] to create a batch of examples. It wil
...
+...     "B-group": 5,
+...     "I-group": 6,
+...     "B-location": 7,
+...     "I-location": 8,
+...     "B-person": 9,
+...     "I-person": 10,
+...     "B-product": 11,
+...     "I-product": 12,
+... }
+```
```
Comment (sgugger) [inline docs/source/en/tasks/token_classification.mdx:257]: This part is framework-agnostic so could go before the split PT/TF.

File: docs/source/en/tasks/token_classification.mdx [hunk 12]
```diff
@@ -239,12 +327,67 @@ Set up an optimizer function, learning rate schedule, and some training hyperpar
 ... )
 ```
 
-Load DistilBERT with [`TFAutoModelForTokenClassification`] along with the number of expected labels:
+Map the expected ids to their labels with `id2label` and `label2id`:
+
+```py
+>>> id2label = {
```
Comment (sgugger) [inline docs/source/en/tasks/token_classification.mdx:333]: Same here, this would avoid this dupe code.

**Review by sgugger — COMMENTED** [review]
Reviewed the last two guides for now, will continue with the other ones later.

File: docs/source/en/tasks/language_modeling.mdx [hunk 13]
```diff
@@ -156,11 +171,11 @@ Apply the `group_texts` function over the entire dataset:
 >>> lm_dataset = tokenized_eli5.map(group_texts, batched=True, num_proc=4)
 ```
 
-For causal language modeling, use [`DataCollatorForLanguageModeling`] to create a batch of examples. It will also *dynamically pad* your text to the length of the longest element in its batch, so they are a uniform length. While it is possible to pad your text in the `tokenizer` function by setting `padding=True`, dynamic padding is more efficient. 
+Now create a batch of examples using [`DataCollatorForLanguageModeling`]. It'll also *dynamically pad* your text to the length of the longest element in its batch (instead of the entire dataset) so they are a uniform length. While it is possible to pad your text in the `tokenizer` function by setting `padding=True`, dynamic padding is more efficient.
```
Comment (sgugger) [inline docs/source/en/tasks/language_modeling.mdx:174]: Same comment as before on the tokenizer here.

File: docs/source/en/tasks/multiple_choice.mdx [hunk 14]
```diff
@@ -79,15 +100,15 @@ The preprocessing function needs to do:
...
 
 ```py
 tokenized_swag = swag.map(preprocess_function, batched=True)
 ```
 
-🤗 Transformers doesn't have a data collator for multiple choice, so you will need to create one. You can adapt the [`DataCollatorWithPadding`] to create a batch of examples for multiple choice. It will also *dynamically pad* your text and labels to the length of the longest element in its batch, so they are a uniform length. While it is possible to pad your text in the `tokenizer` function by setting `padding=True`, dynamic padding is more efficient.
+🤗 Transformers doesn't have a data collator for multiple choice, so you'll need to adapt the [`DataCollatorWithPadding`] to create a batch of examples. It'll also *dynamically pad* your text and labels to the length of the longest element in its batch (instead of the entire dataset) so they are a uniform length. While it is possible to pad your text in the `tokenizer` function by setting `padding=True`, dynamic padding is more efficient.
```
Comment (sgugger) [inline docs/source/en/tasks/multiple_choice.mdx:109]: Same here

File: docs/source/en/tasks/sequence_classification.mdx [hunk 15]
```diff
@@ -91,38 +110,74 @@ Use [`DataCollatorWithPadding`] to create a batch of examples. It will also *dyn
...
 
-If you aren't familiar with fine-tuning a model with the [`Trainer`], take a look at the basic tutorial [here](../training#finetune-with-trainer)!
+If you aren't familiar with finetuning a model with the [`Trainer`], take a look at the basic tutorial [here](../training#train-with-pytorch-trainer)!
 
 </Tip>
+You're ready to start training your model now! Map the expected ids to their labels with `id2label` and `label2id`:
+
+```py
+>>> id2label = {0: "NEGATIVE", 1: "POSITIVE"}
+>>> label2id = {"NEGATIVE": 0, "POSITIVE": 1}
```
Comment (sgugger) [inline docs/source/en/tasks/sequence_classification.mdx:150]: This part should go before we split between frameworks.

File: docs/source/en/tasks/sequence_classification.mdx [hunk 16]
```diff
@@ -66,13 +85,13 @@ Create a preprocessing function to tokenize `text` and truncate sequences to be
...
 
-Use 🤗 Datasets [`~datasets.Dataset.map`] function to apply the preprocessing function over the entire dataset. You can speed up the `map` function by setting `batched=True` to process multiple elements of the dataset at once:
+To apply the preprocessing function over the entire dataset, use 🤗 Datasets [`~datasets.Dataset.map`] function. You can speed up `map` by setting `batched=True` to process multiple elements of the dataset at once:
 
 ```py
 tokenized_imdb = imdb.map(preprocess_function, batched=True)
 ```
 
-Use [`DataCollatorWithPadding`] to create a batch of examples. It will also *dynamically pad* your text to the length of the longest element in its batch, so they are a uniform length. While it is possible to pad your text in the `tokenizer` function by setting `padding=True`, dynamic padding is more efficient.
+Now create a batch of examples using [`DataCollatorWithPadding`]. It'll *dynamically pad* your text to the length of the longest element in its batch (instead of the entire dataset) so they are a uniform length. While it is possible to pad your text in the `tokenizer` function by setting `padding=True`, dynamic padding is more efficient.
```
Comment (sgugger) [inline docs/source/en/tasks/sequence_classification.mdx:94]: Same comment as before here.

File: docs/source/en/tasks/summarization.mdx [hunk 17]
```diff
@@ -84,13 +104,13 @@ The preprocessing function needs to:
...
+To apply the preprocessing function over the entire dataset, use 🤗 Datasets [`~datasets.Dataset.map`] method. You can speed up the `map` function by setting `batched=True` to process multiple elements of the dataset at once:
 
 ```py
 >>> tokenized_billsum = billsum.map(preprocess_function, batched=True)
 ```
 
-Use [`DataCollatorForSeq2Seq`] to create a batch of examples. It will also *dynamically pad* your text and labels to the length of the longest element in its batch, so they are a uniform length. While it is possible to pad your text in the `tokenizer` function by setting `padding=True`, dynamic padding is more efficient.
+Now create a batch of examples using [`DataCollatorForSeq2Seq`]. It will also *dynamically pad* your text and labels to the length of the longest element in its batch (instead of the entire dataset) so they are a uniform length. While it is possible to pad your text in the `tokenizer` function by setting `padding=True`, dynamic padding is more efficient.
```
Comment (sgugger) [inline docs/source/en/tasks/summarization.mdx:113]: Same comment here.

**Review by sgugger — COMMENTED** [review]
Finished reviewing the other tasks. Thanks for putting this all together!

File: docs/source/en/tasks/question_answering.mdx [hunk 18]
```diff
@@ -254,18 +282,349 @@ Configure the model for training with [`compile`](https://keras.io/api/models/mo
...
+>>> question_answerer(question=question, context=context)
+{'score': 0.2058267742395401,
+ 'start': 10,
+ 'end': 95,
+ 'answer': '176 billion parameters and can generate text in 46 languages natural languages and 13'}
+```
+
+You can also manually replicate the results of the `pipeline` if you'd like:
+
+<frameworkcontent>
```
[the remark this answers is not in the document]
Comment (stevhliu) [reply docs/source/en/tasks/question_answering.mdx:560]: I didn't think this section was too long, and it'd be really cool for users to see how they can replicate the pipeline in another way if they wanted to with their own model. I'll keep this here and remove the evaluation section :)

File: docs/source/en/tasks/asr.mdx [hunk 19]
```diff
@@ -99,32 +119,32 @@ The [MInDS-14](https://huggingface.co/datasets/PolyAI/minds14) dataset has a sam
...
 
 ```py
 >>> def prepare_dataset(batch):
 ...     audio = batch["audio"]
 
-...     batch = processor(audio=audio["array"], sampling_rate=audio["sampling_rate"]).input_values[0]
+...     batch["input_values"] = processor(audio["array"], sampling_rate=audio["sampling_rate"]).input_values[0]
 ...     batch["input_length"] = len(batch["input_values"])
 
 ...     batch["labels"] = processor(text=batch["transcription"]).input_ids
```
[the remark this answers is not in the document]
Comment (stevhliu) [reply docs/source/en/tasks/asr.mdx:135]: Super cool! I tried passing it in one call, but I get a `KeyError: 'labels'` (I included a [notebook](https://colab.research.google.com/drive/1Sa0VU9GeABUieYjX2B9SUoK_Fh5JlkxS?usp=sharing)).

[the remark this answers is not in the document]
Comment (sgugger) [reply docs/source/en/tasks/asr.mdx:135]: The first line needs to be
```
batch = processor(audio["array"], sampling_rate=audio["sampling_rate"], text=batch["transcription"])
```
but then it worked for me.

[the remark this answers is not in the document]
Comment (stevhliu) [reply docs/source/en/tasks/asr.mdx]: Ok, this should be fixed now thanks to @sanchit-gandhi ! 👏

File: docs/source/en/tasks/asr.mdx [hunk 20]
```diff
@@ -14,27 +14,44 @@ specific language governing permissions and limitations under the License.
 
 <Youtube id="TksaY_FDgnk"/>
 
-Automatic speech recognition (ASR) converts a speech signal to text. It is an example of a sequence-to-sequence task, going from a sequence of audio inputs to textual outputs. Voice assistants like Siri and Alexa utilize ASR models to assist users.
+Automatic speech recognition (ASR) converts a speech signal to text, going from a sequence of audio inputs to textual outputs. Virtual assistants like Siri and Alexa use ASR models to help users everyday, and there are many other useful user-facing applications like live captioning and note-taking during meetings.
```
Comment (sanchit-gandhi) [inline docs/source/en/tasks/asr.mdx:17]: `textual` sounds a bit funky to me!

```suggestion
Automatic speech recognition (ASR) converts a speech signal to text, mapping a sequence of audio inputs to text outputs. Virtual assistants like Siri and Alexa use ASR models to help users everyday, and there are many other useful user-facing applications like live captioning and note-taking during meetings.
```

File: docs/source/en/tasks/asr.mdx [hunk 21]
```diff
@@ -14,27 +14,44 @@ specific language governing permissions and limitations under the License.
...
 
-This guide will show you how to fine-tune [Wav2Vec2](https://huggingface.co/facebook/wav2vec2-base) on the [MInDS-14](https://huggingface.co/datasets/PolyAI/minds14) dataset to transcribe audio to text.
+This guide will show you how to:
+
+1. Finetune [Wav2Vec2](https://huggingface.co/facebook/wav2vec2-base) on the [MInDS-14](https://huggingface.co/datasets/PolyAI/minds14) to transcribe audio to text.
```
Comment (sanchit-gandhi) [inline docs/source/en/tasks/asr.mdx:21]:
```suggestion
1. Finetune [Wav2Vec2](https://huggingface.co/facebook/wav2vec2-base) on the [MInDS-14](https://huggingface.co/datasets/PolyAI/minds14) dataset to transcribe audio to text.
```

File: docs/source/en/tasks/asr.mdx [hunk 22]
```diff
@@ -14,27 +14,44 @@ specific language governing permissions and limitations under the License.
...
 <Tip>
 
 See the automatic speech recognition [task page](https://huggingface.co/tasks/automatic-speech-recognition) for more information about its associated models, datasets, and metrics.
 
 </Tip>
 
+Before you begin, make sure you have all the necessary libraries installed:
+
+```bash
+pip install transformers datasets evaluate jiwer
```
Comment (sanchit-gandhi) [inline docs/source/en/tasks/asr.mdx:33]: Just checking that this example works without an audio processing library like `soundfile` or `librosa`? Great if it does! Some of these libraries have an annoying installation on Linux

File: docs/source/en/tasks/asr.mdx [hunk 23]
```diff
@@ -14,27 +14,44 @@ specific language governing permissions and limitations under the License.
...
+```py
+>>> from huggingface_hub import notebook_login
+
+>>> notebook_login()
+```
+
 ## Load MInDS-14 dataset
 
-Load the [MInDS-14](https://huggingface.co/datasets/PolyAI/minds14) from the 🤗 Datasets library:
+Start by loading a smaller subset of the [MInDS-14](https://huggingface.co/datasets/PolyAI/minds14) dataset from the 🤗 Datasets library. This'll give you a chance to experiment and make sure everythings works before spending more time training on the full dataset.
```
Comment (sanchit-gandhi) [inline docs/source/en/tasks/asr.mdx:46]:
```suggestion
Start by loading a smaller subset of the [MInDS-14](https://huggingface.co/datasets/PolyAI/minds14) dataset from the 🤗 Datasets library. This'll give you a chance to experiment and make sure everything works before spending more time training on the full dataset.
```

File: docs/source/en/tasks/asr.mdx [hunk 24]
```diff
@@ -99,32 +119,28 @@ The [MInDS-14](https://huggingface.co/datasets/PolyAI/minds14) dataset has a sam
...
 
 ```py
 >>> def prepare_dataset(batch):
 ...     audio = batch["audio"]
-
-...     batch = processor(audio=audio["array"], sampling_rate=audio["sampling_rate"]).input_values[0]
-...     batch["input_length"] = len(batch["input_values"])
-
-...     batch["labels"] = processor(text=batch["transcription"]).input_ids
+...     batch = processor(audio["array"], sampling_rate=audio["sampling_rate"], text=batch["transcription"])
```
Comment (sanchit-gandhi) [inline docs/source/en/tasks/asr.mdx:130]: Nice!

File: docs/source/en/tasks/asr.mdx [hunk 25]
```diff
@@ -212,15 +264,89 @@ At this point, only three steps remain:
...
+The results aren't very good because the model was only trained on a few examples to quickly demonstrate the process. Try finetuning your own model on the full dataset to get better results!
+
+</Tip>
+
+```py
+>>> from transformers import pipeline
+
+>>> transcriber = pipeline("automatic-speech-recognition", model="stevhliu/my_awesome_asr_minds_model")
+>>> transcriber(audio_file)
+{'text': "ALDCBC</s>COCB</s>DCBDBCBRBCDCTCRWKDBK<s></s>KODT</s>C</s>T</s>C</s>OD<s>B</s>BCTOR</s>AKDOBODTCTEDQOATDVDKCDSKL<unk>KEKEPKNPKPADSRDSCLCQ<unk>ED<s>BDXAXDP<unk>KQDPWDKLEPRPKDTDRDPDNFRTADPLDSRDPDKDX<unk>XLXCXLCL</s>LCXLCLDT</s>TBLDLBLTLCB</s>BXTLTBLDTM</s>BLBLCLBLCLDT</s>LXCDTB</s>L</s>T</s>LTDK</s>TDTDTLOKDADPDSDT</s><unk>RDQTDDADSNDN</s>KD</s>T'DPNVPHKDPDRDRKRKD</s>PKDCQS<unk>X<unk>LTAQDXLXCTBCLBTQBCXBDQDTXM<unk>LCXL</s>CXT"}
```
Comment (sanchit-gandhi) [inline docs/source/en/tasks/asr.mdx:315]: Maybe we can train on a few hundred samples to get a somewhat reasonable transcription? Or is the point of this .md purely to demonstrate **how** to fine-tune (rather than with any care to performance), meaning we're happy with mumbo jumbo here?

File: docs/source/en/tasks/audio_classification.mdx [hunk 26]
```diff
@@ -14,27 +14,44 @@ specific language governing permissions and limitations under the License.
 
 <Youtube id="KWwzcmG98Ds"/>
 
-Audio classification assigns a label or class to audio data. It is similar to text classification, except an audio input is continuous and must be discretized, whereas text can be split into tokens. Some practical applications of audio classification include identifying intent, speakers, and even animal species by their sounds.
+Audio classification assigns a label or class to audio data. It is similar to text classification, except an audio input is continuous and must be discretized, whereas text can be split into tokens. Some practical applications of audio classification include identifying speaker intent, language classification, and even animal species by their sounds.
 
-This guide will show you how to fine-tune [Wav2Vec2](https://huggingface.co/facebook/wav2vec2-base) on the [MInDS-14](https://huggingface.co/datasets/PolyAI/minds14) to classify intent.
+This guide will show you how to:
+
+1. Finetune [Wav2Vec2](https://huggingface.co/facebook/wav2vec2-base) on the [MInDS-14](https://huggingface.co/datasets/PolyAI/minds14) to classify speaker intent.
```
Comment (sanchit-gandhi) [inline docs/source/en/tasks/audio_classification.mdx:21]:
```suggestion
1. Finetune [Wav2Vec2](https://huggingface.co/facebook/wav2vec2-base) on the [MInDS-14](https://huggingface.co/datasets/PolyAI/minds14) dataset to classify speaker intent.
```

File: docs/source/en/tasks/audio_classification.mdx [hunk 27]
```diff
@@ -14,27 +14,44 @@ specific language governing permissions and limitations under the License.
 
 <Youtube id="KWwzcmG98Ds"/>
 
-Audio classification assigns a label or class to audio data. It is similar to text classification, except an audio input is continuous and must be discretized, whereas text can be split into tokens. Some practical applications of audio classification include identifying intent, speakers, and even animal species by their sounds.
+Audio classification assigns a label or class to audio data. It is similar to text classification, except an audio input is continuous and must be discretized, whereas text can be split into tokens. Some practical applications of audio classification include identifying speaker intent, language classification, and even animal species by their sounds.
```
Comment (sanchit-gandhi) [inline docs/source/en/tasks/audio_classification.mdx:17]: Actually think that the continuous / discrete statement is a bit confusing? IMO we want to drive the similarity between text and audio classification:
- Text input -> class label output
- Audio input -> class label output

=> same task, different input domain

File: docs/source/en/tasks/audio_classification.mdx [hunk 28]
```diff
@@ -151,25 +199,28 @@ Load Wav2Vec2 with [`AutoModelForAudioClassification`]. Specify the number of la
...
-3. Call [`~Trainer.train`] to fine-tune your model.
+1. Define your training hyperparameters in [`TrainingArguments`]. The only required parameter is `output_dir` which specifies where to save your model. You'll push this model to the Hub by setting `push_to_hub=True` (you need to be signed in to Hugging Face to upload your model). At the end of each epoch, the [`Trainer`] will evaluate the accuracy and save the training checkpoint.
+2. Pass the training arguments to [`Trainer`] along with the model, dataset, tokenizer, data collator, and `compute_metrics` function.
+3. Call [`~Trainer.train`] to finetune your model.
+
 
 ```py
 >>> training_args = TrainingArguments(
-...     output_dir="./results",
+...     output_dir="my_awesome_mind_model",
```
Comment (sanchit-gandhi) [inline docs/source/en/tasks/audio_classification.mdx:211]: Maybe a different name to the ASR one?
```suggestion
...     output_dir="my_awesome_mind_classification_model",
```

File: docs/source/en/tasks/audio_classification.mdx [hunk 29]
```diff
@@ -178,15 +229,89 @@ At this point, only three steps remain:
...
+## Inference
+
+Great, now that you've finetuned a model, you can use it for inference!
+
+Load an audio file you'd like to run inference on. Remember to resample the sampling rate of the audio file to match the sampling rate of the model if you need to!
+
+```py
+>>> from datasets import load_dataset, Audio
+
+>>> dataset = load_dataset("narad/ravdess", split="train")
```
Comment (sanchit-gandhi) [inline docs/source/en/tasks/audio_classification.mdx:261]: Do we want to use this dataset instead of minds for inference?

File: docs/source/en/tasks/asr.mdx [hunk 30]
```diff
@@ -157,17 +173,55 @@ Unlike other data collators, this specific data collator needs to apply a differ
 ...         return batch
 ```
 
-Create a batch of examples and dynamically pad them with `DataCollatorForCTCWithPadding`:
+Now create a batch of examples and dynamically pad them with `DataCollatorForCTCWithPadding`:
```
Comment (sanchit-gandhi) [inline docs/source/en/tasks/asr.mdx:176]: This is potentially a bit misleading since we don't actually batch or pad examples here, but simply instantiate the data collator

**Review by sanchit-gandhi — COMMENTED** [review]
Audio sections look good! Just some minor formatting suggestions from me!

File: docs/source/en/tasks/asr.mdx (hunk 22 above)
Comment (stevhliu) [reply docs/source/en/tasks/asr.mdx:33]: Yeah I was able to run the code without any issues :)

File: docs/source/en/tasks/asr.mdx (hunk 25 above)
Comment (stevhliu) [reply docs/source/en/tasks/asr.mdx:315]: I think I may have run out of memory in Colab when I tried with more samples, but I'll try again with a few hundred samples to see if we can do better than this gibberish.

File: docs/source/en/tasks/audio_classification.mdx (hunk 27 above)
Comment (stevhliu) [reply docs/source/en/tasks/audio_classification.mdx:17]: Good point! 👍

File: docs/source/en/tasks/audio_classification.mdx (hunk 28 above)
Comment (stevhliu) [reply docs/source/en/tasks/audio_classification.mdx:211]: The ASR one is `my_awesome_asr_mind_model` which I think is different enough

File: docs/source/en/tasks/audio_classification.mdx (hunk 29 above)
Comment (stevhliu) [reply docs/source/en/tasks/audio_classification.mdx:261]: Oops, completely wrong dataset! Thanks for catching this :)

File: docs/source/en/tasks/asr.mdx (hunk 25 above)
Comment (sanchit-gandhi) [reply docs/source/en/tasks/asr.mdx:315]: Note that when you download a dataset using `train[:20]` you actually download the entire dataset and then slice the first 20 samples -> so in terms of disk space using a couple of hundred samples will be equivalent to just 20.

For training, you might be getting OOM's because there are very long audio samples in your dataset. The self-attention mechanism memory scales with sequence length squared: $\text{memory} \propto \text{seq len}^2$. So there could be long samples > 20s that are causing OOM's! We usually filter such samples by a 20s max length criteria: 
https://github.com/huggingface/transformers/blob/a0f867430347bcf939f71d186409b9ca138c3b34/examples/pytorch/speech-recognition/run_speech_recognition_ctc.py#L634-L642

Comment (stevhliu) [reply docs/source/en/tasks/asr.mdx]: Finally trained a better model that doesn't return mumbo jumbo and added an important note on making sure the dataset text matches the casing of the tokenizer!

Comment (sanchit-gandhi) [reply docs/source/en/tasks/asr.mdx]: Nice! Glad to hear it's working 🗣

**Review by sgugger — APPROVED** [review]
